### ISSUE TITLE
Do we need to print all 3ˆn ambiguous parses?

### ISSUE THREAD

**Issue by andreasabel** [agda/agda#4713]
My Agda Busy Beaver candidate (shortest code with longest time to error):
```agda
data D : Set where
  c_  : D → D
  c_d : D → D
  _d  : D → D

-- 3ˆn ambiguous parses for test (c cˆn x dˆn d)
test : D → Set
test (c c x d d) = D
-- test (c c c c c c c c c c c x d d d d d d d d d d d) = D
test _ = D
```
The pattern `c ... c x d ... d` is ambiguous.  For the first line, Agda prints 3 possible parsers.
For the commented out line, Agda prints 3ˆ10 ~ 60000 possible parses.

I think we should truncate the list after 10 possible parses.

Also, I don't think we need to compute all these parses eagerly.
https://github.com/agda/agda/blob/d63b57a5db23654cb9ae5b95ff221d8b57020910/src/full/Agda/Syntax/Concrete/Operators.hs#L620-L622
The forcing was added for issue #1129 (f89f8f418360cec340c5cfdd14d2078ede975280), but I'd say it is sufficient to force the final unambiguous result.

**Comment by pnlph** [comment]
I tested the long example.

I can only say that this warning you found is beautiful. If someone wants to use it, it will probably be intentionally. 👨‍🎨

**Comment by andreasabel** [comment]
> If someone wants to use it, it will probably be intentionally.

Yes, one can use the Agda parser as a tree enumerator.

**Comment by pnlph** [comment]
```agda
{-# OPTIONS --cubical #-}
open import Cubical.Hits.S1
3loop : base ≡ base
3loop = loop ∙ loop ∙ loop
```
In this occurrence, the two first results are strangely the same:
```
Don't know how to parse loop ∙ loop ∙ loop. Could mean any one of:
  loop ∙ (loop ∙ loop)
  loop ∙ (loop ∙ loop)
  (loop ∙ loop) ∙ loop
Operators used in the grammar:
  ∙ (infix operator, level 4)   [_∙_ (.../Cubical/Foundations/Id.agda:98,1-4)]      
  ∙ (infixr operator, level 30) [_∙_ (.../Cubical/Foundations/Prelude.agda:117,1-4)]
when scope checking loop ∙ loop ∙ loop
```
When I choose any option
```agda
3loop : base ≡ base
3loop = (loop ∙ loop) ∙ loop
```
Agda still cannot decide from two options that seem the same to the human eye:
```
Don't know how to parse (loop ∙ loop) ∙ loop. Could mean any one
of:
  _∙_ (loop ∙ loop) loop
  _∙_ (loop ∙ loop) loop
```

**Comment by guillaumebrunerie** [comment]
There are somehow two identifiers in scope with the same name (not sure how that happened), one defined in `Cubical/Foundations/Id.agda` and one in `Cubical/Foundations/Prelude.agda`, as you can see from the error message.

But you seem to be using an older version of the cubical library, I can’t reproduce it.
I first get that `Cubical.Hits.S1` doesn’t exist because it has the wrong capitalization (but maybe you are on a Mac). Then I get that `_≡_` is not in scope, and then when I `open import Cubical.Foundations.Prelude` to get `_≡_`, it works.

**Comment by pnlph** [comment]
Thanks, yes this is another case of overloaded identifiers in scope. Sorry, you are right, what I sent was not self-contained. This works:
```agda
open import Cubical.HITs.S1
open import Cubical.Foundations.Prelude
open import Cubical.Data.Int
_ : winding (loop ∙ loop ∙ loop) ≡ pos 3
_ = refl
```

Yes, I am also on a mac so I have issues with capitalization.

Maybe we could raise an additional warning when the ambiguous parses are more than expected because overloading identifiers? It is also true that I did not realize that there were two operators in the grammar with the same name, but I was focused on solving the ambiguous parse.

**Comment by guillaumebrunerie** [comment]
> ```agda
> open import Cubical.HITs.S1
> open import Cubical.Foundations.Prelude
> open import Cubical.Data.Int
> _ : winding (loop ∙ loop ∙ loop) ≡ pos 3
> _ = refl
> ```

This works for me, no parsing issues (with the latest version of the cubical library).

> Maybe we could raise an additional warning when the ambiguous parses are more than expected because overloading identifiers? It is also true that I did not realize that there were two operators in the grammar with the same name, but I was focused on solving the ambiguous parse.

I don’t quite understand how there can be two identifiers with the same name, I thought only constructors/fields can be overloaded. But in any case, this looks like a different issue, maybe you should open a new issue about it.

**Comment by pnlph** [comment]
> I don’t quite understand how there can be two identifiers with the same name, I thought only constructors/fields can be overloaded.

Can you try this code? I think it brings both `_∙_` in scope:
```agda
{-# OPTIONS --cubical --safe #-}
open import Cubical.HITs.S1
open import Cubical.Foundations.Prelude
open import Cubical.Data.Int
open import Cubical.Foundations.Id hiding (_≡_)
_ : winding (loop ∙ loop ∙ loop) ≡ pos 3
_ = refl
```

**Comment by guillaumebrunerie** [comment]
Indeed, this one gives a parse error, but I don’t think that’s an Agda bug, you’re just not supposed to open both `Cubical.Foundations.Id` and `Cubical.Foundations.Prelude` together, as they define the same identifiers on purpose (so that you can choose whether you use path types or identity types).

**Comment by guillaumebrunerie** [comment]
> I don’t quite understand how there can be two identifiers with the same name, I thought only constructors/fields can be overloaded.

Also, I was wrong here. Of course there can be a clash when opening two modules that both define the same identifier.
